### PR TITLE
Balloon enhancement v3

### PR DESCRIPTION
**Pull request by ShirleyFei**
We have two virtio-balloon enhancement features:

- virtio-balloon: add deflate_on_oom support
- virtio-balloon: use fallocate when balloon inflated

### CODE REVIEW

**Comment by ShirleyFei** [comment]
fix the checks

**Comment by liuw** [comment]
I think the code structure looks okay, but I don't know enough of the implementation of virtio-balloon on CLH to make a judgement on the use of `fallocate`.

**Comment by LiHangjing** [comment]
> I think the code structure looks okay, but I don't know enough of the implementation of virtio-balloon on CLH to make a judgement on the use of `fallocate`.

Thank you so much for this review. In qemu, virtio-balloon also uses `fallocate` for hugepages and shmem when balloon inflated. 
The code in here: https://github.com/qemu/qemu/blob/master/softmmu/physmem.c#L3538

**Comment by liuw** [comment]
> > I think the code structure looks okay, but I don't know enough of the implementation of virtio-balloon on CLH to make a judgement on the use of `fallocate`.
> 
> Thank you so much for this review. In qemu, virtio-balloon also uses `fallocate` for hugepages and shmem when balloon inflated.
> The code in here: https://github.com/qemu/qemu/blob/master/softmmu/physmem.c#L3538

Thanks for the pointer. I've checked the link. You code matches what QEMU does.

I was wondering if only setting a feature bit is good enough to enable that feature, so I also checked the virtio balloon implementation in QEMU. It appears the second patch is also correct.  https://git.qemu.org/?p=qemu.git;a=blobdiff;f=hw/virtio/virtio-balloon.c;h=2990f8de5d79b00a740480f37d2885b57186646d;hp=78bc14fc85e267052b59e4a45aa840d5129411b8;hb=e3816255bf4b6377bb405331e2ee0dc14d841b80;hpb=1fa795a853255fcc93e5d3e2a92d161a2ed96eb8

So to the best of my knowledge, this series is correct.

(Closing and reopening this issue to trigger another pipeline run for the failed `cargo fuzz` check)

**Review by liuw — APPROVED** [review]
To be clear: I don't use balloon myself and haven't tested this new feature, but to the best of my knowledge this series matches QEMU's code and doesn't break integration tests.

If don't want this to break in the future, you probably want to add an integration test.

File: virtio-devices/src/balloon.rs [hunk 1]
```diff
@@ -201,6 +203,30 @@ impl BalloonEpollHandler {
 
                 let gpa = (pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT;
                 if let Ok(hva) = mem.get_host_address(GuestAddress(gpa)) {
+                    if queue_index == 0 {
```
Comment (sboeuf) [inline virtio-devices/src/balloon.rs:206]: Don't use the queue index to check if you're in a case of `INFLATE`. Instead, you should do:
```rust
match ev_type {
    INFLATE_QUEUE_EVENT => {
        ...
    }
    _ => {}
}
```

File: virtio-devices/src/balloon.rs [hunk 2]
```diff
@@ -201,6 +203,30 @@ impl BalloonEpollHandler {
...
+                                    host_fd,
+                                    libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
+                                    (offset as u64 + f_offset) as libc::off64_t,
+                                    (1 << VIRTIO_BALLOON_PFN_SHIFT) as libc::off64_t,
+                                )
+                            };
+                            if res != 0 {
+                                return Err(Error::FallocateFail(io::Error::last_os_error()));
+                            }
+                        }
```
Comment (sboeuf) [inline virtio-devices/src/balloon.rs:227]: This whole block can be rewritten:
```rust
                        if let Some(f_off) = region.file_offset() {
                            let offset = hva as usize - region.as_ptr() as usize;
                            let res = unsafe {
                                libc::fallocate64(
                                    f_off.file().as_raw_fd(),
                                    libc::FALLOC_FL_PUNCH_HOLE | libc::FALLOC_FL_KEEP_SIZE,
                                    (offset as u64 + f_off.start()) as libc::off64_t,
                                    (1 << VIRTIO_BALLOON_PFN_SHIFT) as libc::off64_t,
                                )
                            };
                            if res != 0 {
                                return Err(Error::FallocateFail(io::Error::last_os_error()));
                            }
                        }
```

**Review by sboeuf — CHANGES_REQUESTED** [review]
PR looks mostly good, but some code needs to be refactored a bit.

Also, I agree with @liuw it'd be interesting if you can think of a way of triggering OOM in the guest and verify that the balloon has been automatically deflated through an integration test. That's the only way to prevent any regression regarding this feature.

File: virtio-devices/src/balloon.rs (hunk 2 above)
Comment (LiHangjing) [reply virtio-devices/src/balloon.rs:227]: Thank you. I will rewrite this.

Comment (sboeuf) [reply virtio-devices/src/balloon.rs:227]: Cool, let me when the PR is ready for another review!

**Comment by sboeuf** [comment]
Hi @LiHangjing, any update or progress on this PR?

**Comment by LiHangjing** [comment]
> Hi @LiHangjing, any update or progress on this PR?

I have rewrited about patch `virtio-balloon: use fallocate when balloon inflated`. And pull it now. 

Thanks!

**Comment by liuw** [comment]
@LiHangjing unfortunately the pipeline failed. Clippy has identified a better way to clean up the code.

File: virtio-devices/src/balloon.rs [hunk 3]
```diff
@@ -201,6 +203,24 @@ impl BalloonEpollHandler {
 
                 let gpa = (pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT;
                 if let Ok(hva) = mem.get_host_address(GuestAddress(gpa)) {
+                    if let INFLATE_QUEUE_EVENT = ev_type {
+                        let region = mem.find_region(GuestAddress(gpa)).unwrap();
```
Comment (sboeuf) [inline virtio-devices/src/balloon.rs:207]: Please use `map_err()` instead of `unwrap()`.

File: virtio-devices/src/balloon.rs [hunk 4]
```diff
@@ -201,6 +203,24 @@ impl BalloonEpollHandler {
 
                 let gpa = (pfn as u64) << VIRTIO_BALLOON_PFN_SHIFT;
                 if let Ok(hva) = mem.get_host_address(GuestAddress(gpa)) {
+                    if let INFLATE_QUEUE_EVENT = ev_type {
```
Comment (sboeuf) [inline virtio-devices/src/balloon.rs:206]: Why don't you move the entire block inside line 225? We're already checking the `ev_type`, so let's take advantage of it.

File: vmm/src/device_manager.rs [hunk 5]
```diff
@@ -2555,6 +2555,7 @@ impl DeviceManager {
                 virtio_devices::Balloon::new(
                     id.clone(),
                     balloon_config.size,
+                    balloon_config.deflate_on_oom,
```
Comment (sboeuf) [inline vmm/src/device_manager.rs:2558]: Please add some integration testing to validate this feature.

**Review by sboeuf — CHANGES_REQUESTED** [review]

File: vmm/src/device_manager.rs (hunk 5 above)
Comment (ShirleyFei) [reply vmm/src/device_manager.rs:2558]: > Please add some integration testing to validate this feature (if possible).

Ok, sorry for the late reply. We will prepare a corresponding integration testing ASAP. :)

**Comment by ShirleyFei** [comment]
> To be clear: I don't use balloon myself and haven't tested this new feature, but to the best of my knowledge this series matches QEMU's code and doesn't break integration tests.
> 
> If don't want this to break in the future, you probably want to add an integration test.

@liuw @sboeuf Hi all, when I tried to do an integration test, it's a pity that it failed to run. The error is pasted in [2]. Then I do a manual test pasted in [1], by simulating a deflate-on-oom situation:
- Start a 32G guest, and make balloon occupy 10G
- Run a balloon-test.c file to consume memory by continuing to malloc & memset 25G
- Observe the result of 'free -h'
- When more than 22G memory is consumed, we can see balloon begin to deflate and the balloon-test.c continues to run instead of being killed.
- As balloon-test.c finished, we can see the balloon only occupies 6.1G memory now.

So my questions are:
- Is the balloon-test.c convincing enough to validate the deflate-on-oom feature?
- The tests/integration.rs seems to have a bug. Anyway, I will also try to find and fix it. Then I will add a test function into tests/integration.rs just like test_resize() does.

[1] Manual test, see the attached balloon-manual-test.txt
[balloon-manual-test.txt](https://github.com/cloud-hypervisor/cloud-hypervisor/files/6243212/balloon-manual-test.txt)

[2] Integration test
Run `./scripts/run_integration_tests_x86_64.sh`, but find the `No such file or directory` error when running tests/integration.rs. Even run `cargo test --features integration_tests "tests::parallel::test_cpu_topology_142"` failed too, just as the following snapshot:
integration-test-failed

**Comment by sboeuf** [comment]
@ShirleyFei 
>Is the balloon-test.c convincing enough to validate the deflate-on-oom feature?

Yes it is. IMO it's correctly checking that the `deflate-on-oom` feature is supported.

>The tests/integration.rs seems to have a bug. Anyway, I will also try to find and fix it.

What kind of bug? You should always run integration tests with `./scripts/dev_cli.sh tests --integration`.

>Then I will add a test function into tests/integration.rs just like test_resize() does.

Yes that's how it should be done thanks :)

**Comment by sboeuf** [comment]
No progress on this PR, let's close it. Feel free to open a new one when ready.

**Pull request closed without merging**